### PR TITLE
[#89] Symbol Resolution - Initial Implementation

### LINKED ISSUE #89: Symbol Resolution
Add basic support for symbol resolution, i.e. Symbol and Scope.

See also: strumenta/pylasu#5

### PR DESCRIPTION
**Pull request by loradd**
Signed-off-by: Lorenzo Addazi <lorenzo.addazi@strumenta.com>

Closes #89

### CODE REVIEW

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt [hunk 1]
```diff
@@ -65,3 +67,27 @@ fun <N> ReferenceByName<N>.tryToResolve(possibleValue: N?): Boolean where N : Po
         true
     }
 }
+
+interface Symbol : PossiblyNamed
+
+data class Scope(
+    val symbols: MutableMap<String, MutableList<Symbol>> = mutableMapOf(),
```
Comment (ftomassetti) [inline core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt:74]: Maybe we could make symbols private and expose methods to add/read from this table, to ensure it is always used correctly. I.e., we do not want to insert symbols under the wrong key but the key to be always the same as the name of the symbol. Also, we want only to insert symbols with names

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt [hunk 2]
```diff
@@ -65,3 +67,27 @@ fun <N> ReferenceByName<N>.tryToResolve(possibleValue: N?): Boolean where N : Po
...
+    val parent: Scope? = null,
+) {
+    constructor(symbols: MutableList<Symbol>, parent: Scope? = null) : this(
```
Comment (ftomassetti) [inline core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt:77]: here the list could also be immutable. I think List<Symbol> is an ancestor of MutableList

**Review by ftomassetti — APPROVED** [review]
+1

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt [hunk 3]
```diff
@@ -65,3 +67,27 @@ fun <N> ReferenceByName<N>.tryToResolve(possibleValue: N?): Boolean where N : Po
...
+        parent = parent
+    )
+
+    constructor(vararg symbols: Symbol, parent: Scope? = null) : this(
+        symbols = mutableListOf(*symbols),
+        parent = parent
+    )
+}
+
+fun Scope.lookup(symbolName: String, symbolType: Class<*> = Symbol::class.java): Symbol? =
```
[the remark this answers is not in the document]
Comment (loradd) [reply core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt:88]: Thanks, I will add java-specific implementations and replace Class with KClass.

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt [hunk 4]
```diff
@@ -65,3 +67,27 @@ fun <N> ReferenceByName<N>.tryToResolve(possibleValue: N?): Boolean where N : Po
...
 }
+
+interface Symbol : PossiblyNamed
+
+data class Scope(
+    val symbols: MutableMap<String, MutableList<Symbol>> = mutableMapOf(),
+    val parent: Scope? = null,
+) {
+    constructor(symbols: MutableList<Symbol>, parent: Scope? = null) : this(
+        symbols = symbols.groupBy { it.name!! }.mapValues { it.value.toMutableList() }.toMutableMap(),
```
[the remark this answers is not in the document]
Comment (loradd) [reply core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt:78]: 👍

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt (hunk 1 above)
Comment (loradd) [reply core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt:74]: 👍

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt (hunk 2 above)
Comment (loradd) [reply core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt:77]: :+1:

File: emf/metamodel.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
+{
```
Comment (ftomassetti) [inline emf/metamodel.json:1]: I think this one should not be committed (it is probably my fault that this file is generated here by some test)

**Review by ftomassetti — APPROVED** [review]
+1

File: core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt [hunk 6]
```diff
@@ -65,3 +69,33 @@ fun <N> ReferenceByName<N>.tryToResolve(possibleValue: N?): Boolean where N : Po
         true
     }
 }
+
+interface Symbol : PossiblyNamed
+
+@Suppress("DataClassPrivateConstructor")
+data class Scope private constructor(
```
[the remark this answers is not in the document]
Comment (loradd) [reply core/src/main/kotlin/com/strumenta/kolasu/model/Naming.kt]: Right, we can define it as a normal class

File: emf/metamodel.json (hunk 5 above)
Comment (loradd) [reply emf/metamodel.json]: I updated the `emf/.gitignore` file and removed `metamodel.json`

**Review by ftomassetti — APPROVED** [review]
+1 I think you can merge it!

**Pull request merged**